node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,234 @@
+      case _ =>
+        Sync[F].raiseError[Vector[TransactionInfo]](
+          new Exception(
+            "It is not possible that user deploy got more than 3 reports(precharge, user, refund)."
+          )
+        )
+    }
+  }
+
+  def getTransaction(blockHash: String): F[Seq[TransactionInfo]] =
+    for {
+      report <- blockReportAPI.blockReport(blockHash, forceReplay = false)
+      ts <- report match {
+             case Right(b) =>
+               for {
+                 udts <- b.deploys.toList.traverse(d => {
+                          d.report.zipWithIndex.toList.traverse {
+                            case (report, index) => findUserDeployTransaction(d, report, index)
+                          }
+                        })
+                 userTransactions = udts.flatten.flatten
+                 systemDeployTransactions = b.systemDeploys.flatMap { s =>
+                   // system doesn't get precharge and refund , so it would always get one
+                   val transactions = findTransactions(s.report.head)
+                   s.systemDeploy.get.systemDeploy.value match {
+                     case SlashSystemDeployDataProto(_, _) =>
+                       transactions.map(
+                         t => TransactionInfo(t, SlashingDeploy(blockHash))
+                       )
+                     case CloseBlockSystemDeployDataProto() =>
+                       transactions.map(t => TransactionInfo(t, CloseBlock(blockHash)))
+                   }
+                 }
+                 transactions = userTransactions ++ systemDeployTransactions
+               } yield transactions
+
+             case Left(_) => Seq.empty.pure
+           }
+    } yield ts
+
+  private def findTransactions(report: SingleReport): Vector[Transaction] = {
+    val transactions = report.events.foldLeft(Vector.empty[Transaction]) {
+      case (acc, reportProto) =>
+        reportProto.report.value match {
+          case ReportCommProto(Some(consume), produces) =>
+            consume.channels.headOption
+              .flatMap { channel =>
+                if (channel == transferUnforgeable) {
+                  // based on RevVault.rho:line188 transfer method
+                  // _transferTemplate!(ownRevAddress, *purse, revAddress, amount, *authKey, *ret2)
+                  val produce        = produces.head
+                  val fromAddr       = produce.data.get.pars.head.exprs.head.getGString
+                  val toAddr         = produce.data.get.pars(2).exprs.head.getGString
+                  val amount         = produce.data.get.pars(3).exprs.head.getGInt
+                  val retUnforgeable = produce.data.get.pars(5)
+                  val transaction    = Transaction(fromAddr, toAddr, amount, retUnforgeable, None)
+                  Some(acc :+ transaction)
+                } else None
+              }
+              .getOrElse(acc)
+
+          case _ => acc
+        }
```

Comment:
This kind of collecting can be simplify without fold and with cleaner separation of what is collected and how is processed.
```scala
// First the collecting part
report.events
  .map(_.report.value)
  .collect {
    case ReportCommProto(Some(consume), produces) =>
      consume.channels.headOption.map((_, produces))
  }
  .flatten
  .flatMap {
    // And then when we have all what we need, comes processing part
    case (channel, produces) =>
      if (channel == transferUnforgeable) {
        // ...
        Seq(transaction)
      } else Seq.empty
  }
```
Or even step further.
```scala
report.events
  .map(_.report.value)
  .collect {
    case ReportCommProto(Some(consume), produces) =>
      consume.channels.headOption.map((_, produces))
  }
  .flatten
  .collect { case (channel, produces) if channel == transferUnforgeable => produces }
  .map { produces =>
    // ...
    transaction
  }
```

Reply:
Isn't that going to iterate several times?